hydpy/models/rconc/rconc_control.py

```diff
@@ -362,67 +388,45 @@ def set_triangle_uh(self, tb: float, tp: Optional[float] = None) -> None:
-            uh1[:] = sh1j - sh1j_1
-
-            # sum should be equal to one but better normalize
-            uh1[:] = uh1 / numpy.sum(uh1)
-
-    def set_gr_uh2_uh(self, x4: float, beta: float = 2.5) -> None:
-        """Calculate and set the ordinates of a unit hydrograph as used in GR model
-        (UH2)
+        self.shape, self.values = self._det_gr_uh(x4=x4, beta=beta, left=True)
+        self.subpars.pars.model.sequences.logs.quh.shape = self.shape
+
+    def set_gr_uh2(self, x4: float, beta: float = 2.5) -> None:
+        """Calculate and set the ordinates of a unit hydrograph as done by GR4J for
+        modifying direct runoff."""
+        shape_left, values_left = self._det_gr_uh(x4=x4, beta=beta, left=True)
+        shape_right, values_right = self._det_gr_uh(x4=x4, beta=beta, left=False)
+        self.shape = shape_left + shape_right - 1
+        self.subpars.pars.model.sequences.logs.quh.shape = self.shape
+        self.values[:] = 0.0
+        self.values[:shape_left] = values_left / 2.0
+        self.values[-shape_right:] += values_right / 2.0
+
+    def _det_gr_uh(self, x4: float, beta: float, left: bool) -> tuple[int, VectorFloat]:
+        """Determine the shape and the ordinates of a GR4J-like (part of a) unit
+        hydrograph.
+
+        `left` means the rising limb and corresponds to "uh1" and the left part of
+        "uh2".
         """
-        # TODO: refactor
-        quh = self.subpars.pars.model.sequences.logs.quh
-        if x4 <= 1.0:
-            index = numpy.arange(1, 3)
-        else:
-            index = numpy.arange(1, numpy.ceil(x4 * 2) + 1)
+        if (x4 <= 0.5) or (left and x4 <= 1.0):
+            return 1, numpy.ones(1, dtype=float)
 
-        nmb_uhs = len(index)
-        sh2j = numpy.zeros(shape=nmb_uhs)
-        sh2j_1 = numpy.zeros(shape=nmb_uhs)
-
-        for idx in range(nmb_uhs):
-            if index[idx] <= x4:
-                sh2j[idx] = 0.5 * (index[idx] / x4) ** beta
-            elif x4 < index[idx] < 2.0 * x4:
-                sh2j[idx] = 1.0 - 0.5 * (2.0 - index[idx] / x4) ** beta
-            else:
-                sh2j[idx] = 1
-
-            if index[idx] - 1 <= x4:
-                sh2j_1[idx] = 0.5 * ((index[idx] - 1) / x4) ** beta
-            elif x4 < index[idx] - 1 < 2.0 * x4:
-                sh2j_1[idx] = 1.0 - 0.5 * (2.0 - (index[idx] - 1) / x4) ** beta
-            else:
-                # sh2j_1[idx] = 1
-                assert False, "Please check gr_uh2 algorithm, line has been removed"
-
-        self.shape = len(index)
-        quh.shape = len(index)
-        uh2 = self.values
-        uh2[:] = sh2j - sh2j_1
-        uh2[:] = uh2 / numpy.sum(uh2)
+        if left:
+            ts = numpy.arange(0.0, x4 + 1.0)
+        else:
+            ts = numpy.arange(2.0 * x4 - numpy.ceil(x4) + 1.0, -1.0, -1.0)[::-1]
+        totals = numpy.empty(len(ts), dtype=float)
+        totals[1:-1] = (ts[1:-1] / x4) ** beta
+        totals[0], totals[-1] = 0.0, 1.0
```

Comment:
This might be a little clearer and a tiny bit more efficient:

```python
        if left:
            ts = numpy.arange(1.0, x4)
        else:
            ts = numpy.arange(2.0 * x4 - numpy.ceil(x4), 0.0, -1.0)[::-1]
        totals = numpy.empty(len(ts) + 2, dtype=float)
        totals[1:-1] = (ts / x4) ** beta
        totals[0], totals[-1] = 0.0, 1.0
```